File: backend/src/main/java/com/sankalai/service/ProfileService.java
package com.sankalai.service;

import com.sankalai.dto.*;
import com.sankalai.entity.*;
import com.sankalai.repository.*;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

@Service
public class ProfileService {

    private final UserRepository userRepository;
    private final UserStatsRepository userStatsRepository;
    private final AchievementRepository achievementRepository;
    private final HomeService homeService;
    private final FocusService focusService;
    private final BlockingService blockingService;

    public ProfileService(UserRepository userRepository, UserStatsRepository userStatsRepository,
            AchievementRepository achievementRepository, HomeService homeService,
            FocusService focusService, BlockingService blockingService) {
        this.userRepository = userRepository;
        this.userStatsRepository = userStatsRepository;
        this.achievementRepository = achievementRepository;
        this.homeService = homeService;
        this.focusService = focusService;
        this.blockingService = blockingService;
    }

    // Achievement definitions matching frontend
    private static final List<AchievementDefinition> ACHIEVEMENT_DEFINITIONS = List.of(
            // Common Achievements
            new AchievementDefinition("first-session", "First Steps", "Complete your first focus session", "🎯", Achievement.Rarity.COMMON, Achievement.Category.SESSIONS, 1),
            new AchievementDefinition("early-bird", "Early Bird", "Complete a session before 8 AM", "🌅", Achievement.Rarity.COMMON, Achievement.Category.SPECIAL, 1),
            new AchievementDefinition("night-owl", "Night Owl", "Complete a session after 10 PM", "🦉", Achievement.Rarity.COMMON, Achievement.Category.SPECIAL, 1),
            new AchievementDefinition("streak-3", "Getting Started", "Maintain a 3-day streak", "🔥", Achievement.Rarity.COMMON, Achievement.Category.STREAK, 3),
            new AchievementDefinition("sessions-10", "Dedicated Learner", "Complete 10 focus sessions", "📚", Achievement.Rarity.COMMON, Achievement.Category.SESSIONS, 10),
            // Rare Achievements
            new AchievementDefinition("streak-7", "Week Warrior", "Maintain a 7-day streak", "⚡", Achievement.Rarity.RARE, Achievement.Category.STREAK, 7),
            new AchievementDefinition("deep-work-first", "Deep Diver", "Complete your first 90-minute deep work session", "🌊", Achievement.Rarity.RARE, Achievement.Category.FOCUS, 1),
            new AchievementDefinition("sessions-50", "Consistent Scholar", "Complete 50 focus sessions", "🎓", Achievement.Rarity.RARE, Achievement.Category.SESSIONS, 50),
            new AchievementDefinition("level-10", "Rising Star", "Reach Level 10", "⭐", Achievement.Rarity.RARE, Achievement.Category.LEVEL, 10),
            new AchievementDefinition("blocking-master", "Discipline Master", "Complete 20 blocking sessions without breaking", "🛡️", Achievement.Rarity.RARE, Achievement.Category.BLOCKING, 20),
            // Epic Achievements
            new AchievementDefinition("streak-30", "Monthly Master", "Maintain a 30-day streak", "🏆", Achievement.Rarity.EPIC, Achievement.Category.STREAK, 30),
            new AchievementDefinition("sessions-100", "Centurion", "Complete 100 focus sessions", "💯", Achievement.Rarity.EPIC, Achievement.Category.SESSIONS, 100),
            new AchievementDefinition("deep-work-10", "Deep Work Expert", "Complete 10 deep work sessions", "🧠", Achievement.Rarity.EPIC, Achievement.Category.FOCUS, 10),
            new AchievementDefinition("level-50", "Adept Scholar", "Reach Level 50", "🌟", Achievement.Rarity.EPIC, Achievement.Category.LEVEL, 50),
            new AchievementDefinition("perfect-week", "Perfect Week", "Complete sessions 7 days in a row with 90+ focus score", "💎", Achievement.Rarity.EPIC, Achievement.Category.SPECIAL, 7),
            // Legendary Achievements
            new AchievementDefinition("streak-100", "Unstoppable", "Maintain a 100-day streak", "👑", Achievement.Rarity.LEGENDARY, Achievement.Category.STREAK, 100),
            new AchievementDefinition("sessions-500", "Grand Master", "Complete 500 focus sessions", "🎖️", Achievement.Rarity.LEGENDARY, Achievement.Category.SESSIONS, 500),
            new AchievementDefinition("level-100", "Enlightened", "Reach the maximum level (100)", "✨", Achievement.Rarity.LEGENDARY, Achievement.Category.LEVEL, 100),
            new AchievementDefinition("deep-work-50", "Flow State Master", "Complete 50 deep work sessions", "🌌", Achievement.Rarity.LEGENDARY, Achievement.Category.FOCUS, 50),
            new AchievementDefinition("perfect-score", "Perfectionist", "Achieve 100 focus score in 10 sessions", "🏅", Achievement.Rarity.LEGENDARY, Achievement.Category.SPECIAL, 10)
    );

    /**
     * Get user profile
     */
    public UserProfileDTO getUserProfile(String userId) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("User not found"));

        UserStats userStats = homeService.getUserStats(userId);
        var focusStats = focusService.getFocusStats(userId);
        var achievements = getUserAchievements(userId);

        long unlockedCount = achievements.stream().filter(AchievementDTO::unlocked).count();

        return new UserProfileDTO(
                userId,
                user.getUsername(),
                user.getEmail(),
                user.getAvatar(),
                user.getExam(),
                user.getExamName(),
                user.getSubjects(),
                user.getEmailVerified(),
                userStats.getCurrentLevel(),
                userStats.getCurrentFocusPoints(),
                userStats.getTotalSessions(),
                userStats.getTotalMinutes() / 60,
                userStats.getCurrentStreak(),
                ACHIEVEMENT_DEFINITIONS.size(),
                (int) unlockedCount,
                0, // globalRank (placeholder)
                user.getJoinedAt(),
                user.getLastActive());
    }

    /**
     * Get user achievements
     */
    public List<AchievementDTO> getUserAchievements(String userId) {
        // Get user's unlocked achievements from database
        List<Achievement> userAchievements = achievementRepository.findByUser_UserId(userId);

        Map<String, LocalDateTime> unlockedIds = userAchievements.stream()
                .filter(a -> a.getUnlocked())
                .collect(Collectors.toMap(
                        Achievement::getAchievementId,
                        Achievement::getUnlockedAt
                ));

        // Get user stats for progress calculation
        UserStats userStats = homeService.getUserStats(userId);
        var focusStats = focusService.getFocusStats(userId);
        var blockingStats = blockingService.getBlockingStats(userId);
        var sessions = focusService.getUserSessions(userId, 100);

        // Map achievements with progress
        return ACHIEVEMENT_DEFINITIONS.stream()
                .map(def -> {
                    boolean unlocked = unlockedIds.containsKey(def.id());
                    int currentValue = getCurrentValue(def, userStats, focusStats, blockingStats, sessions);
                    int progress = unlocked ? 100 : Math.min(100, (currentValue * 100) / def.requirement());

                    return new AchievementDTO(
                            def.id(),
                            def.title(),
                            def.description(),
                            def.icon(),
                            def.rarity().name().toLowerCase(),
                            def.category().name().toLowerCase(),
                            unlocked,
                            unlockedIds.get(def.id()),
                            progress,
                            def.requirement(),
                            currentValue);
                })
                .collect(Collectors.toList());
    }

    /**
     * Get current value for achievement progress
     */
    private int getCurrentValue(
            AchievementDefinition achievement,
            UserStats userStats,
            FocusStatsDTO focusStats,
            com.sankalai.dto.BlockingDTO.BlockingStatsDTO blockingStats,
            List<com.sankalai.entity.FocusSession> sessions
    ) {
        return switch (achievement.id()) {
            case "first-session", "sessions-10", "sessions-50", "sessions-100", "sessions-500" ->
                    userStats.getTotalSessions();
            case "streak-3", "streak-7", "streak-30", "streak-100" ->
                    userStats.getCurrentStreak();
            case "level-10", "level-50", "level-100" ->
                    userStats.getCurrentLevel();
            case "deep-work-first", "deep-work-10", "deep-work-50" ->
                    focusStats.totalDeepWorkSessions();
            case "blocking-master" ->
                    blockingStats.completedSessions();
            case "early-bird" ->
                    (int) sessions.stream()
                            .filter(s -> s.getStartTime().getHour() < 8)
                            .count();
            case "night-owl" ->
                    (int) sessions.stream()
                            .filter(s -> s.getStartTime().getHour() >= 22)
                            .count();
            case "perfect-week" -> {
                LocalDateTime weekAgo = LocalDateTime.now().minusDays(7);
                long count = sessions.stream()
                        .filter(s -> s.getStartTime().isAfter(weekAgo) && s.getFocusScore() >= 90)
                        .count();
                yield (int) count;
            }
            case "perfect-score" ->
                    (int) sessions.stream()
                            .filter(s -> s.getFocusScore() != null && s.getFocusScore() == 100)
                            .count();
            default -> 0;
        };
    }

    /**
     * Check and unlock achievements
     */
    @Transactional
    public void checkAndUnlockAchievements(String userId) {
        List<AchievementDTO> achievements = getUserAchievements(userId);

        List<AchievementDTO> toUnlock = achievements.stream()
                .filter(a -> !a.unlocked() && a.currentValue() >= a.requirement())
                .toList();

        if (toUnlock.isEmpty()) return;

        User user = userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("User not found"));

        for (AchievementDTO achievement : toUnlock) {
            Achievement userAchievement = Achievement.builder()
                    .user(user)
                    .achievementId(achievement.id())
                    .title(achievement.title())
                    .description(achievement.description())
                    .icon(achievement.icon())
                    .rarity(Achievement.Rarity.valueOf(achievement.rarity().toUpperCase()))
                    .category(Achievement.Category.valueOf(achievement.category().toUpperCase()))
                    .unlocked(true)
                    .unlockedAt(LocalDateTime.now())
                    .progress(100)
                    .requirement(achievement.requirement())
                    .currentValue(achievement.currentValue())
                    .build();

            achievementRepository.save(userAchievement);
        }

        // Award bonus FP for unlocking achievements
        int bonusFP = toUnlock.size() * 50; // 50 FP per achievement
        if (bonusFP > 0) {
            homeService.awardRewards(userId, bonusFP, 0);
        }
    }

    /**
     * Get leaderboard (top users by level and FP)
     */
    public List<LeaderboardEntryDTO> getLeaderboard(int limitCount) {
        // This would require a more complex query in production
        // For now, return empty array
        // In production, you'd need to:
        // 1. Create a separate leaderboard collection
        // 2. Update it periodically with Cloud Functions
        // 3. Query it here
        return List.of();
    }

    /**
     * Update user profile
     */
    @Transactional
    public void updateUserProfile(String userId, ProfileUpdateRequest updates) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("User not found"));

        if (updates.username() != null) {
            user.setUsername(updates.username());
        }
        if (updates.avatar() != null) {
            user.setAvatar(updates.avatar());
        }
        if (updates.exam() != null) {
            user.setExam(updates.exam());
        }
        if (updates.examName() != null) {
            user.setExamName(updates.examName());
        }
        if (updates.subjects() != null) {
            user.setSubjects(updates.subjects());
        }

        userRepository.save(user);
    }

    // Record types
    record AchievementDefinition(
            String id,
            String title,
            String description,
            String icon,
            Achievement.Rarity rarity,
            Achievement.Category category,
            int requirement
    ) {}
}

File: frontend/src/services/profileService.ts
// Profile Service - User profile and achievements management with backend API integration
import { apiCall } from '../config/api';

// ═══════════════════════════════════════════════════════════
// TYPES
// ═══════════════════════════════════════════════════════════

export interface Achievement {
  id: string;
  title: string;
  description: string;
  icon: string;
  rarity: 'common' | 'rare' | 'epic' | 'legendary';
  category: 'sessions' | 'streak' | 'focus' | 'blocking' | 'level' | 'special';
  unlocked: boolean;
  unlockedAt?: Date;
  progress?: number;        // 0-100 for locked achievements
  requirement: number;      // What's needed to unlock
  currentValue: number;     // User's current progress
}

export interface UserProfile {
  userId: string;
  username: string;
  email: string;
  avatar: string;
  
  // Stats Summary
  level: number;
  focusPoints: number;
  totalSessions: number;
  totalHours: number;
  currentStreak: number;
  
  // Achievements
  totalAchievements: number;
  unlockedAchievements: number;
  
  // Rank
  globalRank?: number;
  
  // Timestamps
  joinedAt: Date;
  lastActive: Date;
}

// ═══════════════════════════════════════════════════════════
// ACHIEVEMENT DEFINITIONS
// ═══════════════════════════════════════════════════════════

export const ACHIEVEMENT_DEFINITIONS: Omit<Achievement, 'unlocked' | 'unlockedAt' | 'progress' | 'currentValue'>[] = [
  // Common Achievements
  {
    id: 'first-session',
    title: 'First Steps',
    description: 'Complete your first focus session',
    icon: '🎯',
    rarity: 'common',
    category: 'sessions',
    requirement: 1,
  },
  {
    id: 'early-bird',
    title: 'Early Bird',
    description: 'Complete a session before 8 AM',
    icon: '🌅',
    rarity: 'common',
    category: 'special',
    requirement: 1,
  },
  {
    id: 'night-owl',
    title: 'Night Owl',
    description: 'Complete a session after 10 PM',
    icon: '🦉',
    rarity: 'common',
    category: 'special',
    requirement: 1,
  },
  {
    id: 'streak-3',
    title: 'Getting Started',
    description: 'Maintain a 3-day streak',
    icon: '🔥',
    rarity: 'common',
    category: 'streak',
    requirement: 3,
  },
  {
    id: 'sessions-10',
    title: 'Dedicated Learner',
    description: 'Complete 10 focus sessions',
    icon: '📚',
    rarity: 'common',
    category: 'sessions',
    requirement: 10,
  },

  // Rare Achievements
  {
    id: 'streak-7',
    title: 'Week Warrior',
    description: 'Maintain a 7-day streak',
    icon: '⚡',
    rarity: 'rare',
    category: 'streak',
    requirement: 7,
  },
  {
    id: 'deep-work-first',
    title: 'Deep Diver',
    description: 'Complete your first 90-minute deep work session',
    icon: '🌊',
    rarity: 'rare',
    category: 'focus',
    requirement: 1,
  },
  {
    id: 'sessions-50',
    title: 'Consistent Scholar',
    description: 'Complete 50 focus sessions',
    icon: '🎓',
    rarity: 'rare',
    category: 'sessions',
    requirement: 50,
  },
  {
    id: 'level-10',
    title: 'Rising Star',
    description: 'Reach Level 10',
    icon: '⭐',
    rarity: 'rare',
    category: 'level',
    requirement: 10,
  },
  {
    id: 'blocking-master',
    title: 'Discipline Master',
    description: 'Complete 20 blocking sessions without breaking',
    icon: '🛡️',
    rarity: 'rare',
    category: 'blocking',
    requirement: 20,
  },

  // Epic Achievements
  {
    id: 'streak-30',
    title: 'Monthly Master',
    description: 'Maintain a 30-day streak',
    icon: '🏆',
    rarity: 'epic',
    category: 'streak',
    requirement: 30,
  },
  {
    id: 'sessions-100',
    title: 'Centurion',
    description: 'Complete 100 focus sessions',
    icon: '💯',
    rarity: 'epic',
    category: 'sessions',
    requirement: 100,
  },
  {
    id: 'deep-work-10',
    title: 'Deep Work Expert',
    description: 'Complete 10 deep work sessions',
    icon: '🧠',
    rarity: 'epic',
    category: 'focus',
    requirement: 10,
  },
  {
    id: 'level-50',
    title: 'Adept Scholar',
    description: 'Reach Level 50',
    icon: '🌟',
    rarity: 'epic',
    category: 'level',
    requirement: 50,
  },
  {
    id: 'perfect-week',
    title: 'Perfect Week',
    description: 'Complete sessions 7 days in a row with 90+ focus score',
    icon: '💎',
    rarity: 'epic',
    category: 'special',
    requirement: 7,
  },

  // Legendary Achievements
  {
    id: 'streak-100',
    title: 'Unstoppable',
    description: 'Maintain a 100-day streak',
    icon: '👑',
    rarity: 'legendary',
    category: 'streak',
    requirement: 100,
  },
  {
    id: 'sessions-500',
    title: 'Grand Master',
    description: 'Complete 500 focus sessions',
    icon: '🎖️',
    rarity: 'legendary',
    category: 'sessions',
    requirement: 500,
  },
  {
    id: 'level-100',
    title: 'Enlightened',
    description: 'Reach the maximum level (100)',
    icon: '✨',
    rarity: 'legendary',
    category: 'level',
    requirement: 100,
  },
  {
    id: 'deep-work-50',
    title: 'Flow State Master',
    description: 'Complete 50 deep work sessions',
    icon: '🌌',
    rarity: 'legendary',
    category: 'focus',
    requirement: 50,
  },
  {
    id: 'perfect-score',
    title: 'Perfectionist',
    description: 'Achieve 100 focus score in 10 sessions',
    icon: '🏅',
    rarity: 'legendary',
    category: 'special',
    requirement: 10,
  },
];

// ═══════════════════════════════════════════════════════════
// PROFILE SERVICE CLASS
// ═══════════════════════════════════════════════════════════

class ProfileService {
  /**
   * Get user profile from backend
   */
  async getUserProfile(userId: string): Promise<UserProfile> {
    try {
      const response = await apiCall('/profile', 'GET') as any;
      
      return {
        userId: response.userId,
        username: response.username,
        email: response.email,
        avatar: response.avatar,
        level: response.level,
        focusPoints: response.focusPoints,
        totalSessions: response.totalSessions,
        totalHours: response.totalHours,
        currentStreak: response.currentStreak,
        totalAchievements: response.totalAchievements,
        unlockedAchievements: response.unlockedAchievements,
        globalRank: response.globalRank,
        joinedAt: new Date(response.joinedAt),
        lastActive: new Date(response.lastActive),
      };
    } catch (error) {
      console.error('Error getting user profile:', error);
      throw error;
    }
  }

  /**
   * Get user achievements from backend
   */
  async getUserAchievements(userId: string): Promise<Achievement[]> {
    try {
      const response = await apiCall('/profile/achievements', 'GET') as any;
      
      return response.map((a: any) => ({
        id: a.id,
        title: a.title,
        description: a.description,
        icon: a.icon,
        rarity: a.rarity,
        category: a.category,
        unlocked: a.unlocked,
        unlockedAt: a.unlockedAt ? new Date(a.unlockedAt) : undefined,
        progress: a.progress,
        requirement: a.requirement,
        currentValue: a.currentValue,
      }));
    } catch (error) {
      console.error('Error getting user achievements:', error);
      return [];
    }
  }

  /**
   * Check and unlock achievements via backend
   */
  async checkAndUnlockAchievements(userId: string): Promise<void> {
    try {
      await apiCall('/profile/achievements/check', 'POST');
    } catch (error) {
      console.error('Error checking achievements:', error);
      throw error;
    }
  }

  /**
   * Get leaderboard from backend
   */
  async getLeaderboard(limitCount: number = 50): Promise<{
    userId: string;
    username: string;
    avatar: string;
    level: number;
    focusPoints: number;
    rank: number;
  }[]> {
    try {
      const response = await apiCall(`/profile/leaderboard?limit=${limitCount}`, 'GET') as any;
      
      return response.map((entry: any) => ({
        userId: entry.userId,
        username: entry.username,
        avatar: entry.avatar,
        level: entry.level,
        focusPoints: entry.focusPoints,
        rank: entry.rank,
      }));
    } catch (error) {
      console.error('Error getting leaderboard:', error);
      return [];
    }
  }

  /**
   * Update user profile via backend
   */
  async updateUserProfile(
    userId: string,
    updates: {
      username?: string;
      avatar?: string;
    }
  ): Promise<void> {
    try {
      await apiCall('/profile', 'PUT', updates);
    } catch (error) {
      console.error('Error updating user profile:', error);
      throw error;
    }
  }

  // Placeholder methods for backward compatibility
  private getCurrentValue(
    achievement: Omit<Achievement, 'unlocked' | 'unlockedAt' | 'progress' | 'currentValue'>,
    userStats: any,
    focusStats: any,
    blockingStats: any,
    sessions: any[]
  ): number {
    // Progress is calculated on the backend
    return 0;
  }
}

export const profileService = new ProfileService();
